File: go/money_test.go
package main

import (
    s "tdd/stocks"
    "testing"
    "reflect"

)

var bank s.Bank

func init() {
    bank = s.NewBank()
    bank.AddExchangeRate("EUR", "USD", 1.2)
    bank.AddExchangeRate("USD", "KRW", 1100)
}

func TestMultiplicationInDollars(t *testing.T) {
    fiveDollars := s.NewMoney(5, "USD")
    actualResult := fiveDollars.Times(2)
    expectedResult := s.NewMoney(10, "USD")
    assertEqual(t, expectedResult, actualResult)
}

func TestMultiplicationInEuros(t *testing.T) {
    tenEuros := s.NewMoney(10, "EUR")
    actualResult := tenEuros.Times(2)
    expectedResult := s.NewMoney(20, "EUR")
    assertEqual(t, expectedResult, actualResult)
}

func TestDivision(t *testing.T) {
    originalMoney := s.NewMoney(4002, "KRW")
    actualResult := originalMoney.Divide(4)
    expectedResult := s.NewMoney(1000.5, "KRW")
    assertEqual(t, expectedResult, actualResult)
}

func TestAddition(t *testing.T) {
    var portfolio s.Portfolio

    fiveDollars := s.NewMoney(5, "USD")
    tenDollars := s.NewMoney(10, "USD")
    fifteenDollars := s.NewMoney(15, "USD")

    portfolio = portfolio.Add(fiveDollars)
    portfolio = portfolio.Add(tenDollars)
    portfolioInDollars, err := portfolio.Evaluate(bank, "USD")

    assertNil(t, err)
    assertEqual(t, fifteenDollars, *portfolioInDollars)
}

func TestAdditionOfDollarsAndEuros(t *testing.T) {
    var portfolio s.Portfolio

    fiveDollars := s.NewMoney(5, "USD")
    tenEuros := s.NewMoney(10, "EUR")

    portfolio = portfolio.Add(fiveDollars)
    portfolio = portfolio.Add(tenEuros)

    expectedValue := s.NewMoney(17, "USD")
    actualValue, err := portfolio.Evaluate(bank, "USD")

    assertNil(t, err)
    assertEqual(t, expectedValue, *actualValue)
}

func TestAdditionOfDollarsAndWons(t *testing.T) {
    var portfolio s.Portfolio

    oneDollar := s.NewMoney(1, "USD")
    elevenHundredWon := s.NewMoney(1100, "KRW")

    portfolio = portfolio.Add(oneDollar)
    portfolio = portfolio.Add(elevenHundredWon)

    expectedValue := s.NewMoney(2200, "KRW")
    actualValue, err := portfolio.Evaluate(bank, "KRW")

    assertNil(t, err)
    assertEqual(t, expectedValue, *actualValue)
}

func TestAdditionWithMultipleMissingExcahngeRates(t *testing.T) {
    var portfolio s.Portfolio

    oneDollar := s.NewMoney(1, "USD")
    oneEuro := s.NewMoney(1, "EUR")
    oneWon := s.NewMoney(1, "KRW")

    portfolio = portfolio.Add(oneDollar)
    portfolio = portfolio.Add(oneEuro)
    portfolio = portfolio.Add(oneWon)

    expectedErrorMessage := 
        "Missing exchange rate(s):[USD->Kalganid,EUR->Kalganid,KRW->Kalganid,]"
    value, actualError := portfolio.Evaluate(bank, "Kalganid")
    
    assertNil(t, value)
    assertEqual(t, expectedErrorMessage, actualError.Error())
}

func TestConversion(t *testing.T) {
    bank := s.NewBank()
    bank.AddExchangeRate("EUR", "USD", 1.2)
    tenEuros := s.NewMoney(10, "EUR")
    actualConvertedMoney, err := bank.Convert(tenEuros, "USD")
    assertNil(t, err)
    assertEqual(t, s.NewMoney(12, "USD"), *actualConvertedMoney)
}

func TestConversionWithMissingExchangeRate(t *testing.T) {
    bank := s.NewBank()
    tenEuros := s.NewMoney(10, "EUR")
    actualConvertedMoney, err := bank.Convert(tenEuros, "Kalganid")
    if actualConvertedMoney != nil {
        t.Errorf("Expected money to be nill, found: [%+v]", actualConvertedMoney)
    }
    assertNil(t, actualConvertedMoney)
    assertEqual(t, "EUR->Kalganid", err.Error())
}

func assertNil(t *testing.T, actual interface{}) {
    if actual != nil && !reflect.ValueOf(actual).IsNil() {
        t.Errorf("Expected error to be nil, found [%+v]", actual)
    }
}

func assertEqual(t *testing.T, expected interface{}, actual interface{}) {
    if expected != actual {
        t.Errorf("Expected %+v, got %+v", expected, actual)
    }
}

File: php/tests/MoneyTest.php
<?php

namespace App\Tests;

use PHPUnit\Framework\TestCase;
use App\Money;
use App\Portfolio;
use App\Bank;

class MoneyTest extends TestCase
{
    private Bank $bank;
     
    public function setUp(): void 
    {
        parent::setUp();
        $this->bank = new Bank();
        $this->bank->addExchangeRate("EUR", "USD", 1.2);
        $this->bank->addExchangeRate("USD", "KRW", 1100);
    }

    public function testMultiplicationInDollars(): void
    {
        $fiveDollars = new Money(5, "USD");
        $tenDollars = $fiveDollars->times(2);
        $this->assertTrue($tenDollars->isEqual($fiveDollars->times(2)));
    }

    public function testMultiplicationInEuros(): void
    {
        $tenEuros = new Money(10, "EUR");
        $twentyEuros = $tenEuros->times(2);
        $this->assertTrue($twentyEuros->isEqual($tenEuros->times(2)));
    }

    public function testDivision(): void
    {
        $originalMoney = new Money(4002, "KRW");
        $actualMoneyAfterDivision = $originalMoney->divide(4);
        $expectedMoneyAfterDivision = new Money(1000.5, "KRW");
        $this->assertEquals($expectedMoneyAfterDivision->getAmount(), $actualMoneyAfterDivision->getAmount());
        $this->assertEquals($expectedMoneyAfterDivision->getCurrency(), $actualMoneyAfterDivision->getCurrency());
    }

    public function testAddition(): void
    {
        $fiveDollars = new Money(5, "USD");
        $tenDollars = new Money(10, "USD");
        $fifteenDollars = new Money(15, "USD");
        $portfolio = new Portfolio();
        $portfolio->add($fiveDollars, $tenDollars);
        $this->assertTrue($fifteenDollars->isEqual($portfolio->evaluate($this->bank, "USD")));
    }

    public function testConversion(): void
    {
        $bank = new Bank();
        $bank->addExchangeRate("EUR", "USD", 1.2);
        $tenEuros = new Money(10, "EUR");
        $this->assertEquals($bank->convert($tenEuros, "USD"), new Money(12, "USD"));
        $this->assertEquals($this->bank->convert($tenEuros, "USD"), new Money(12, "USD"));
        $this->bank->addExchangeRate("EUR", "USD", 1.3);
        $this->assertEquals($this->bank->convert($tenEuros, "USD"), new Money(13, "USD"));
    }   

    public function testAdditionOfDollarsAndEuros(): void
    {
        $fiveDollars = new Money(5, "USD");
        $tenEuros = new Money(10, "EUR");
        $portfolio = new Portfolio();
        $portfolio->add($fiveDollars, $tenEuros);
        $expectedValue = new Money(17, "USD");
        $actualValue = $portfolio->evaluate($this->bank, "USD");
        $this->assertTrue($expectedValue->isEqual($actualValue));
    }

    public function testAdditionOfDollarsAndWons(): void
    {
        $oneDollar = new Money(1, "USD");
        $elevenHundredWons = new Money(1100, "KRW");
        $portfolio = new Portfolio();
        $portfolio->add($oneDollar, $elevenHundredWons);
        $expectedValue = new Money(2200, "KRW");
        $actualValue = $portfolio->evaluate($this->bank, "KRW");
        $this->assertTrue($expectedValue->isEqual($actualValue));
    }

    public function testAdditionWithMultipleMissingExchangeRates(): void
    {
        $oneDollar = new Money(1, "USD");
        $oneEuro = new Money(1, "EUR");
        $oneWon = new Money(1, "KRW");
        $portfolio = new Portfolio();
        $portfolio->add($oneDollar, $oneEuro, $oneWon);
        $this->expectException(\Exception::class);
        $this->expectExceptionMessage("Missing exchange rate(s):[Kalganid->Kalganid]");
        $portfolio->evaluate($this->bank, "Kalganid");
    }

    public function testWhatIsTheConversionRateFromEURToUSD(): void
    {
        $tenEuros = new Money(10, "EUR");
        $this->assertEquals($this->bank->convert($tenEuros, "USD"), new Money(12, "USD"));
    }
}
